### PR TITLE
Always use XDG directories

### LINKED ISSUE #136: Sheldon should follow XDG Base Directory spec even if environment variables are not defined
From https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html:
> $XDG_DATA_HOME defines the base directory relative to which user-specific data files should be stored. If $XDG_DATA_HOME is either not set or empty, a default equal to $HOME/.local/share should be used.
>
> $XDG_CONFIG_HOME defines the base directory relative to which user-specific configuration files should be stored. If $XDG_CONFIG_HOME is either not set or empty, a default equal to $HOME/.config should be used.

Most apps respect and follow this behavior, so sheldon should default to `$HOME/.local/share/sheldon` and `$HOME/.config/sheldon`. Perhaps the best way to have forward compatibility with this change is to detect both the xdg directories and `~/.sheldon` but give a warning when `~/.sheldon` is loaded.

### PR DESCRIPTION
**Pull request by jcgruenhage**
Resolves #136

### CODE REVIEW

File: src/lock/file.rs [hunk 1]
```diff
@@ -58,6 +58,10 @@ impl LockedConfig {
         P: AsRef<Path>,
     {
         let path = path.as_ref();
+        if let Some(parent) = path.parent() {
+            std::fs::create_dir_all(parent)
+                .with_context(s!("failed to write locked config to `{}`, because the parent directory couldn't be created", path.display()))?;
+        }
```
Comment (jcgruenhage) [inline src/lock/file.rs:66]: this is new, because without it, sheldon complained to me that it couldn't lock the plugins. Not sure if this is the best place, but it's at least a good enough place.

File: src/cli/mod.rs [hunk 2]
```diff
@@ -125,48 +125,33 @@ impl Opt {
...
-
-        let (config_pre, data_pre) = if using_xdg {
-            (
-                xdg_config_user
-                    .unwrap_or_else(|| home.join(".config"))
-                    .join("sheldon"),
-                xdg_data_user
-                    .unwrap_or_else(|| home.join(".local/share"))
-                    .join("sheldon"),
+        let project_dirs = match directories::ProjectDirs::from("rs", "cli", "sheldon").ok_or_else(|| {
```
Comment (rossmacarthur) [inline src/cli/mod.rs]: Based on the docs for this function I think the following makes more sense

```rust
directories::ProjectDirs::from("rs.cli", "", "sheldon")
```

File: src/cli/mod.rs [hunk 3]
```diff
@@ -125,48 +125,33 @@ impl Opt {
             no_color: color.is_no_color(),
         };
 
-        let home = match home.or_else(home::home_dir).ok_or_else(|| {
+        let base_dirs = match directories::BaseDirs::new().ok_or_else(|| {
```
Comment (rossmacarthur) [inline src/cli/mod.rs]: I think there's some opportunity for code reuse here, perhaps something like this?

```rust
        let log_no_home = || {
            let err = anyhow!(
                "failed to determine the current user's home directory, \
                 try setting the $HOME environment variable"
            );
            log_error(output.no_color, Color::Red, "error", &err);
        };
        let base_dirs = match directories::BaseDirs::new() {
            Some(base_dirs) => base_dirs,
            None => {
                log_no_home();
                process::exit(1);
            }
        };
        let project_dirs = match directories::ProjectDirs::from("rs.cli", "", "sheldon") {
            Some(project_dirs) => project_dirs,
            None => {
                log_no_home();
                process::exit(1);
            }
        };
```

File: src/cli/mod.rs [hunk 4]
```diff
@@ -125,48 +125,33 @@ impl Opt {
...
+                log_error(output.no_color, Color::Red, "error", &err);
+                process::exit(1);
+            }
         };
 
-        let config_dir = config_dir.unwrap_or(config_pre);
-        let data_dir = data_dir.unwrap_or(data_pre);
+        let home = home.unwrap_or(base_dirs.home_dir().to_path_buf());
+        let config_dir = config_dir.unwrap_or(project_dirs.config_dir().to_path_buf());
+        let data_dir = data_dir.unwrap_or(project_dirs.data_dir().to_path_buf());
```
Comment (rossmacarthur) [inline src/cli/mod.rs]: Let's use `.unwrap_or_else` to prevent unnecessary clones in the case that an option is set.

```rust
let home = home.unwrap_or_else(|| base_dirs.home_dir().to_owned());
let config_dir = config_dir.unwrap_or_else(|| project_dirs.config_dir().to_owned());
let data_dir = data_dir.unwrap_or_else(|| project_dirs.data_dir().to_owned());
```

File: src/lock/file.rs (hunk 1 above)
Comment (rossmacarthur) [reply src/lock/file.rs:66]: This seems fine. The calling function already gives some context so I think you just make the context here

```rust
s!("failed to create parent directory for `{}`, path.display())
```

**Review by rossmacarthur — COMMENTED** [review]
Thanks for the contribution. I'm happy with this approach and using the `directories` crate for this. I think lets not worry too much about migration plans and rather release this in v0.7.0 and document the changes. We can perhaps write a function that checks the old directory and logs a warning but I'm happy for that to be in a second PR.

File: src/lock/file.rs (hunk 1 above)
Comment (jcgruenhage) [reply src/lock/file.rs:66]: Okay, adapted the context.

File: src/cli/mod.rs (hunk 2 above)
Comment (jcgruenhage) [reply src/cli/mod.rs]: Re-reading the documentation:

> An empty string can be passed if no organization should be used

Seems you're right here, will adapt

**Comment by jcgruenhage** [comment]
With the most recent version of this patch, the documentation was also adjusted to explain the new behaviour. @rossmacarthur how does the changelog work? Where would I put a section about the changes here?

**Comment by jcgruenhage** [comment]
Thought about migration logic again, and came to the conclusion that it's less pain than I thought previously. Migrating data from old places to new places is implemented now, although I haven't tested it extensively yet.

File: src/cli/mod.rs [hunk 5]
```diff
@@ -190,6 +191,30 @@ impl Opt {
     }
 }
 
+fn get_and_migrate_dir<P: AsRef<Path> + PartialEq>(
```
Comment (rossmacarthur) [inline src/cli/mod.rs:194]: I would prefer not to actually move the directories. I think this could be a bit invasive. I'd rather just log a warning and link to the release notes where we can document what options people have.

if we do add something like this, I think it should be its own subcommand and I would prefer that be done in a separate PR.

File: docs/src/Getting-started.md [hunk 6]
```diff
@@ -16,16 +16,17 @@ or
 sheldon init --shell zsh
 ```
 
-This will create `plugins.toml` under `~/.sheldon` or, if defined,
-`$XDG_CONFIG_HOME/sheldon`. You can either edit this file directly or use the
-provided command line interface to add or remove plugins.
+This will create `plugins.toml` in the appropriate config home of your
+platform. On most Linux systems this will be `~/.config/sheldon/`. You can
+either edit this file directly or use the provided command line interface to
+add or remove plugins.
```
Comment (rossmacarthur) [inline docs/src/Getting-started.md:22]: Let's also document the default macOS directory, which looks to be

```
~/Library/Application Support/rs.cli.sheldon
```

**Review by rossmacarthur — COMMENTED** [review]
Thanks for updating the documentation. I have some concerns about the migration code but otherwise I'm happy with these changes.

File: src/cli/mod.rs (hunk 5 above)
Comment (jcgruenhage) [reply src/cli/mod.rs:194]: We're only moving the directory if they were using the default dir before and are using the default dir afterwards. Moving is less invasive than breaking IMO, but you're right that doing so silently might be problematic. What do you think about logging a warning and moving? As for a separate subcommand: I think that'd just make it even more complicated, and we should probably avoid that.

**Comment by rossmacarthur** [comment]
Closing this because it was resolved in c3f87dde3017e98845378148acc4cd2268572a7a
